crates/sui-indexer-alt-jsonrpc/src/api/coin.rs

```diff
@@ -0,0 +1,349 @@
+            }
+        }
+
+        // Construct the basic query first to filter by owner, not deleted and newest rows.
+        let mut query = candidates
+            .select(candidates!(
+                object_id,
+                cp_sequence_number,
+                coin_balance_bucket
+            ))
+            .left_join(
+                newer.on(candidates!(object_id)
+                    .eq(newer!(object_id))
+                    .and(candidates!(cp_sequence_number).lt(newer!(cp_sequence_number)))),
+            )
+            .filter(newer!(object_id).is_null())
+            .filter(candidates!(owner_kind).eq(StoredCoinOwnerKind::Fastpath))
+            .filter(candidates!(owner_id).eq(owner.to_vec()))
+            .into_boxed();
+
+        // If the coin type is specified, we filter by it.
+        if let Some(coin_type_tag) = coin_type_tag {
+            let serialized_coin_type =
+                bcs::to_bytes(&coin_type_tag).context("Failed to serialize coin type tag")?;
+            query = query.filter(candidates!(coin_type).eq(serialized_coin_type));
+        }
+
+        // If the cursor is specified, we filter by it.
+        if let Some(Cursor(BalanceCursor {
+            object_id,
+            cp_sequence_number,
+            coin_balance_bucket,
+        })) = page.cursor
+        {
+            query = query.filter(sql::<Bool>(&format!(
+                r#"ROW(candidates.coin_balance_bucket, candidates.cp_sequence_number, candidates.object_id)
+                 < ROW({coin_balance_bucket}, {cp_sequence_number}, '\x{object_id}'::bytea)"#,
+                object_id = hex::encode(object_id),
+            )));
+        }
+
+        // Finally we order by coin_balance_bucket and break ties by object_id.
+        query = query
+            .order_by(candidates!(coin_balance_bucket).desc())
+            .then_order_by(candidates!(cp_sequence_number).desc())
+            .then_order_by(candidates!(object_id).desc())
+            .limit(limit + 1);
+
+        #[derive(Queryable, Debug, Serialize, Deserialize)]
+        #[diesel(table_name = coin_balance_buckets)]
+        struct IdCpBalance {
+            object_id: Vec<u8>,
+            cp_sequence_number: i64,
+            coin_balance_bucket: Option<i16>,
+        }
+
+        let mut buckets: Vec<IdCpBalance> =
+            conn.results(query).await.context("Failed to query coins")?;
+
+        // Now gather pagination info.
+        let has_next_page = buckets.len() > limit as usize;
+        if has_next_page {
+            buckets.truncate(limit as usize);
+        }
+
+        let next_cursor = buckets
+            .last()
+            .map(|bucket| {
+                let object_id =
+                    ObjectID::from_bytes(&bucket.object_id).context("Failed to parse object id")?;
+                let balance_bucket = bucket
+                    .coin_balance_bucket
+                    .ok_or_else(|| anyhow::anyhow!("Coin balance bucket is null"))?;
```

Comment:
You probably won't need it in this case (if you set `assume_not_null` above) but I recently learned that `anyhow`'s `context` works on optional values as well, so you can do:

```suggestion
                    .context("Coin balance bucket is null")?;
```